### ISSUE TITLE
[docs] Automatically generate the documentation for the exported modules

### ISSUE THREAD

**Issue by oliviertassinari** [mui-org/material-ui-x#1213]
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

The data grid is not like the other core components. It has a very large API surface, with many modules and types (not only components).

## Motivation 🔦

The documentation is critical to the success of the product. This issue is an extension of #1211, and probably even more important.

In the main repository, we have started to face this pain but didn't try to solve it properly. For the few modules we export that are not components, we manually maintain the API with a somewhat consistent markdown format. This includes hooks, custom functions. There isn't a lot.

With the data grid, it currently hard to understand, what's available, as well as confusing. I was looking into https://material-ui.com/components/data-grid/filtering/#custom-filter-operator and it took me a while to under how it was working.

cc @mui-org/maintainers 

## Solution 🧩

One proposal for the solution (for the whole mui-org):

1. We replace the API navigation

<img width="252" alt="Screenshot 2021-03-13 at 19 32 23" src="https://user-images.githubusercontent.com/3165635/111040342-e3340300-8432-11eb-81e0-52ed32855675.png">

with a page that links all the others. We can even structure the data, with headers per category. This would allow scaling to more modules. I think that we reached a point where there are simply too many links to have a side nav like this keep having sense.

2. We create an API extractor, to convert the types into a json structure the documentation can render. I made a POC in this direction on #1212. It seems to work, it's fast. A [preview](https://deploy-preview-1212--material-ui-x.netlify.app/).

3. We build a React component to render json data. It has to be fast. We might have a lot of information to render

4. We increase the coverage. The first possible tradeoff is to render the API only where appropriate. For instance, we could replace this manually authored markdown https://material-ui.com/components/app-bar/#usescrolltrigger-options-trigger with it. We could also include the description of [GridFilterOperator](https://deploy-preview-1212--material-ui-x.netlify.app/interfaces/gridfilteroperator) on this page https://material-ui.com/components/data-grid/filtering/#custom-filter-operator. Now, if we move forward with 5, 4. might not be as relevant. We could link the other pages instead.

5. We consider documenting all the public API. In the POC with typedocs, #1212, I didn't check that it only generates the types for the public modules. We would need to double-check. My main concern here is scalability. How much time does it take to build? How many new pages would Google need to crawl (there is a budget)? Maybe we should bypass React and our components here. Use plain old HTML + CSS.

## Benchmark 👀

- https://www.telerik.com/kendo-react-ui/components/grid/api/
- https://lodash.com/docs/4.17.15#pickBy
- https://ag-grid.com/javascript-grid/grid-api/
- https://stripe.com/docs/js/payment_request/events
- https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html
- https://react-restart.github.io/hooks/api/useFocusManager
- https://react-select.com/props
- https://react-table.tanstack.com/docs/api/useColumnOrder (I have heard multiple times that the docs of react-table is far far from great).

**Comment by eps1lon** [comment]
I don't follow why we need to change the component API documentation. It seems to me you just want to auto-generate API docs for something other than components. Could we focus the issue on that?

Rendering everything onto a single page is not viable. I wouldn't want to load the full API just to get the API of a single component. We could solve this with a combination of different techniques but why not use something that's been working for the web for decades now: multiple pages.

> How many new pages would Google need to crawl (there is a budget)?

Don't try to find problems before they exists. It's trivial to find potential problems but not trivial to solve these hypothetical problems.

And lastly: I don't understand what this "benchmark" is for. These are vastly different types of APIs and I don't see an analysis attached. How did you pick these pages? Why did you pick them? What pages did you omit? Why did you omit them? What problems do these pages solve? What problems do they not solve? How good is the solution? 

> I have heard multiple times that the docs of react-table is far far from great

What is this supposed to mean? Why are they not "good"? Who said it was not "good"? When where these statements made.

Overall, I don't think this issue is fleshed out enough to start a constructive discussion. In the end, why do we need to change the material-ui.com docs if the problem is not with material-ui.com?

**Comment by oliviertassinari** [comment]
> I don't follow why we need to change the component API documentation. It seems to me you just want to auto-generate API docs for something other than components. Could we focus the issue on that?

@eps1lon You are right. The two problems are not directly related. My thought was the following: assuming we solve the generation for something other than components, wouldn't it mean that we x2 the number of pages we need to generate? With twice as many pages, does the linear navigation in the side nav still make sense? Happy to focus on the main issue, and leave this one for later. It could simply be considered as a constraint to not ignore.

> Rendering everything onto a single page is not viable. I wouldn't want to load the full API just to get the API of a single component. We could solve this with a combination of different techniques but why not use something that's been working for the web for decades now: multiple pages.

If you are making a reference to lodash, I can't agree more! I personally hate the UX of the docs. async went on the same path. I get that it's simpler to search on the page, but with a tool like Algolia, why.
I should have been clearer. In 1. I was only proposing the change the link structure, but to keep individual URLs for each module

> Don't try to find problems before they exists. It's trivial to find potential problems but not trivial to solve these hypothetical problems.

You are right, the last time this was a real problem for a product, we around one million pages crawled per day. We will never be anywhere near this scale at Material-UI. We currently have 10k crawls a day. So we are 2 orders of magnitude behind, and with content that doesn't change at a high frequency.

> I don't understand what this "benchmark" is for.

My objective with the benchmark here is to get a sample of what other people have tried as solutions. My hope is it can give us **inspiration** for finding a solution that is intrinsically great. In practice, if we draft a proposed solution, we can see it already implemented somewhere and we see that it feels great once implemented, this would already, from my perspective, be a big win for the use case of the benchmark.
I don't think that we should expect anything more. Representativity and biases in the dataset aren't relevant. It's significantly skewed by what I have linked.

> What is this supposed to mean? Why are they not "good"? Who said it was not "good"? When where these statements made.

I'm sharing what I have heard on the documentation. It would suggest that we shouldn't do what they are doing 😁. In practice, does it help us move the solution forward? No, not really.

> Overall, I don't think this issue is fleshed out enough to start a constructive discussion.

Fair. The solution I have sketched is very high-level and doesn't get, at all, into the details, where a lot of the complexity is. I hope that the **description of the problem** was enough so if somebody wants to give a shot to the problem, share a different proposal for the solution. 

> In the end, why do we need to change the material-ui.com docs if the problem is not with material-ui.com?

I didn't understand this question.

**Comment by oliviertassinari** [comment]
One quick exploration https://github.com/mui-org/material-ui-x/pull/288

**Comment by eps1lon** [comment]
> we can see it already implemented somewhere and we see that it feels great once implemented,

That is not how that works. How do you know what problem a given implementation is trying to solve? How do you know it got solved?

**Comment by oliviertassinari** [comment]
@eps1lon I personally think that the problem a benchmarked solution has tried to solve is secondary.
I was taking it in the angle of: Does this things they have build solve the very specific problem we have?

In the same way that when somebody purchase something, e.g a car, he will evaluate if the solution fits his problem, even if the solution might have been optimized for a different persona. 🤷‍♂️

**Comment by oliviertassinari** [comment]
The POC done by @m4theushw #1529.

The challenges/steps I have seen so far:

- [ ] Update the monorepo infrastructure of MUI X to depend on HEAD (not v4). Until we do, we can test the solution in any of the repo and bring it to the other once it works. I personally have a preference for starting with the core because it's harder (force to make the real tradeoffs, not wasting time on oversimplistic ones)
- [ ] Extract the API information from the types and generate a rich normalized JSON structure.
- [ ] Solve the Next.js performance bottleneck (to go from 100 to 1,000 API pages). We are currently creating a new page for each exported API. It's slow to build. Instead, we could/should build one dynamic API page, and generate all the API pages, statically.
- [ ] Create a new full-page render to handle different types. So far, we have done one for the component type. We could have one for hooks (if it ever make sense, or simplify functions), etc.
- [ ] Create in-page renderng blocks, in the same way we have MDX like components: 

https://github.com/mui-org/material-ui/blob/36c1e08520f8add7bdc341825d44f50d59ceab83/docs/src/pages/components/alert/alert.md#L14

This could be helpful when we want to inline API right inside the main docs pages, the alternative is to reference the API reference section.

- [ ] Fix the side nav, linking each API reference in the side can't scale, it breaks the UX.
